pkg/syncproto/sync_proto.go

```diff
@@ -12,6 +12,141 @@
+//	|<-----------------------|
+//	|                        |
+//	|     SyncStatus(InSync) |
+//	|<-----------------------|
+//	|                        |
+//	|               KVs  * n |
+//	|<-----------------------|
+//	|                        |
+//
+// Wire format
+//
+// The protocol uses gob to encode messages.  Each message is wrapped in an Envelope
+// struct to simplify decoding.
+//
+// Key/value pairs are encoded as SerializedUpdate objects.  These contain the KV pair
+// along with the Syncer metadata about the update (such as its revision and update type).
+// The key and value are encoded to the libcalico-go "default" encoding, as used when
+// storing data in, for example, etcd.  I.e. the gob struct contains string and []byte
+// fields to hold the key and value, respectively.  Doing this has a some advantages:
+//
+//     (1) It avoids any subtle incompatibility between our datamodel and gob.
+//
+//     (2) It removes the need to register all our datatypes with the gob en/decoder.
+//
+//     (3) It re-uses known-good serialization code with known semantics around
+//         data-model upgrade.
+//
+//     (4) It allows us to do the serialization of each KV pair once and send it to
+//         all listening clients.  (Doing this in gob is not easy because the gob
+//         connection is stateful.)
+//
+// Upgrading the datamodel
+//
+// Some care needs to be taken when upgrading Felix and Typha to ensure that datamodel
+// changes are correctly handled.
+//
+// Since Typha parses resources from the datamodel and then serializes them again,
+// Typha can only pass through resources (and fields) that were present in the version of
+// libcalico-go that it was compiled against.  Similarly, Felix can only parse resources
+// and fields that were present in the version of libcalico-go that it was compiled against.
+//
+// In the common case, where a new field is added to the datamodel:
+//
+//     - If a new Felix connects to an old Typha then Typha will strip the new field
+//       at parse-time and pass the object through to Felix.  Hence Felix will behave
+//       as if the field wasn't present.  As long as the field was added in a back-compatible
+//       way, Felix should default to its old behaviour and the overall outcome will be
+//       that new Felix will behave as f it was an old Felix.
+//
+//     - If an old Felix connects to a new Typha, then Typha will pass through the new
+//       field to Felix but Felix will strp it out when it parses the update.
+//
+// Where a whole new resource is added:
+//
+//     - If a new Felix connects to an old Typha then Typha will ignore the new resource
+//       so it is important that Felix is engineered to allow for missing resources in
+//       that case.
+//
+//     - If an old Felix connects to a new Typha then Typha will send the resource
+//       but the old Felix will fail to parse it.  In that case, the Typha client code
+//       drops the KV pair and logs and error.
```

Comment:
"the Typha client code" -> "Felix's Typha client code" (wasn't totally clear on my first parsing of this that the Typha client code was referring to part of Felix)